### DISCUSSION TITLE
update line plot dynamically

### DISCUSSION THREAD

**Discussion by zillionare** [h2oai/wave#1477] [Q&A]
All examples I found was plot line graph with initial values, however, I need to update the graph as time goes. According to the doc, it as easy as append data to `data` property of plot card:
this is what I have did:

```py
async def render_content(q: Q):
    q.page["content"] = ui.plot_card(
        box="content",
        title="Line, groups",
        data=data("date series assets", 250, rows=[(0, "0", 0.5), (1, "0", 1)]),
        plot=ui.plot(
            [ui.mark(type="line", x="=date", y="=assets", color="=series", y_min=0)]
        ),
    )
```

this will give me the following screen:

<img width="1153" alt="image" src="https://user-images.githubusercontent.com/7206696/171846884-d76178e6-78f5-4256-b1f0-dc124613d907.png">

However, I'm failed to update it with:
```
async def on_backtest_progress(q):
    for i in range(100):
        await asyncio.sleep(1)
        try:
            q.page["content"].data[i] = [i, str(i // 2), random.random() * i]
            await q.page.save()
        except Exception as e:
            pass
```
the plot never change. What I got from console log:

```
2022/06/03 19:23:28 * /98e5953c-abac-4f57-ab50-7a56fc732520 {"d":[{"k":"content data 97","v":[97,"48",2.562474580531558]}]}
2022/06/03 19:23:29 * /98e5953c-abac-4f57-ab50-7a56fc732520 {"d":[{"k":"content data 98","v":[98,"49",67.13149453163314]}]}
2022/06/03 19:23:30 * /98e5953c-abac-4f57-ab50-7a56fc732520 {"d":[{"k":"content data 99","v":[99,"49",21.670791363664435]}]}
```

by the way, the document usually refers `data` property as `plot_data`(check https://wave.h2o.ai/docs/buffers), however, it actually `data`. Guess this is a version update issue.

**Comment by zillionare** [comment]
A workaround is redraw the card every time, but still a more efficient way is better.

**Comment by zillionare** [comment]
found why.

if the buffer is defined as an array, then the initial size counters. If the initial rows contains 2 elements, then only the very first element will be updated; if the initial rows is empty, then it will not draw any graph at all.

circular buffer doesn't have this issue.

**Comment by lo5** [comment]
The array and circular buffers have a fixed number of rows, as indicated here: https://wave.h2o.ai/docs/buffers#types-of-buffers

The map buffer can have a variable number of rows.

The three buffer types exist to improve latency for high-frequency updates (several times per second). If you don't plan to do that, it's better to just pass `rows=` or `columns=` with `pack=True`: https://wave.h2o.ai/docs/buffers/#packed-buffers
